5.7 regression, closure not accepted when function with same signature is

With this code:

```swift
struct Foo<T> {
    init(_: (T) -> Void) {}
}

protocol Bar {}

enum Baz: Bar {
    case someCase(Int)
}

let _: Foo<Bar> = Foo<Bar> { (a: Bar) -> Void in
    switch a {
    case let .someCase(value) as Baz:
        print(value)
    default:
        break
    }
}
```

Compiling with Swift 5.6 works fine. With Swift 5.7 (Xcode 14b6) I get:

```
% swiftc bar.swift
bar.swift:11:19: error: type of expression is ambiguous without more context
let _: Foo<Bar> = Foo<Bar> { (a: Bar) -> Void in
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

With Swift main @ https://github.com/apple/swift/commit/5bfef4fc4bfa6e424384a39533ceea3744a613cc I get:

```
/tmp/bar.swift:13:15: error: type 'any Bar' has no member 'someCase'
    case let .someCase(value) as Baz:
             ~^~~~~~~~
```

Which appears to hint at the issue. Changing the closure to a function with the same signature sidesteps the issue:

```swift
struct Foo<T> {
    init(_: (T) -> Void) {}
}

protocol Bar {}

enum Baz: Bar {
    case someCase(Int)
}

func b(a: Bar) -> Void {
    switch a {
    case let .someCase(value) as Baz:
        print(value)
    default:
        break
    }
}

let _: Foo<Bar> = Foo<Bar>(b)
```

--- a/lib/Sema/CSGen.cpp
+++ b/lib/Sema/CSGen.cpp
@@ -2721,6 +2721,18 @@ namespace {
               locator.withPathElement(LocatorPathElt::PatternMatch(pattern)));
 
           baseType = parentType;
+          // Perform member lookup into the external pattern metatype. e.g.
+          // `case let .test(tuple) as Test`.
+        } else if (externalPatternType) {
+          Type externalMetaType = MetatypeType::get(externalPatternType);
+
+          CS.addValueMemberConstraint(
+              externalMetaType, enumPattern->getName(), memberType, CurDC,
+              functionRefKind, {},
+              CS.getConstraintLocator(locator,
+                                      LocatorPathElt::PatternMatch(pattern)));
+
+          baseType = externalPatternType;
         } else {
           // Use the pattern type for member lookup.
           CS.addUnresolvedValueMemberConstraint(

--- a/test/Constraints/issue-60806.swift
+++ b/test/Constraints/issue-60806.swift
@@ -1,0 +1,31 @@
+// RUN: %target-swift-frontend -dump-ast %s | %FileCheck %s
+
+// https://github.com/apple/swift/issues/60806
+
+struct Foo<T> {
+  init(_: (T) -> Void) {}
+}
+
+protocol Bar {}
+
+enum Baz: Bar {
+  case someCase(Int)
+}
+
+enum NonBarBaz {
+  case someCase(Int)
+}
+
+let _: Foo<Bar> = Foo<Bar> { (a: Bar) -> Void in
+  switch a {
+  // CHECK: (pattern_is type='Bar' value_cast Baz
+  // CHECK-NEXT: (pattern_enum_element type='Baz' Baz.someCase
+  case let .someCase(value) as Baz:
+    print(value)
+  // expected-warning@-1 {{cast from 'any Bar' to unrelated type 'NonBarBaz' always fails}}
+  case let .someCase(value) as NonBarBaz:
+    print(value)
+  default:
+      break
+  }
+}

